runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -59,16 +250,175 @@ extern "C" {
+            break;
+        default:
+            Assert_VM_unreachable();
+    }
+
+    I_32 gprArgCount = 0;
+    I_32 fprArgCount = 0;
+    // We have 3 GPRs (R1 to R3) and 4 FPRs (F0, F2, F4, F6) available for parameter passing.
+    const I_32 maxGPRArgCount = 3;
+    const I_32 maxFPRArgCount = 4;
+
+    // Loop through the arguments and count how many instructions you'll need to copy data from registers to memory, and how many
+    // bytes of space you'll need to allocate in thunk and new stack frame to store the instructions + parameter info.
+    for (I_32 i = 0; i < lastSigIdx; i++) {
+        I_32 argSizeInBytes = sigArray[i].sizeInByte;
+        switch (sigArray[i].type) {
+            case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT64:
+                if (gprArgCount < maxGPRArgCount) {
+                    numInstructions++;
+                }
+                gprArgCount++;
+                break;
+            case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+            case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+                if (fprArgCount < maxFPRArgCount) {
+                    numInstructions++;
+                }
+                fprArgCount++;
+                break;
+            default:
+                Assert_VM_unreachable();
+        }
+        if (fprArgCount + gprArgCount >= 7)
+            break;
+
+    }
+
+    // The thunk will frame will be 256 bytes in size. Since parameters will be stored in the caller frame, offsetFromR4
+    // must account for the frame size of the thunk.
+    const I_32 defaultStackPointerBias = 2048;
+    const I_32 offsetToParamArea = 128;
+    const I_32 calleeFrameSize = 256;
+    const I_32 offsetFromR4 = defaultStackPointerBias + offsetToParamArea + calleeFrameSize;
+    I_8 *thunkMem = NULL;
+
+    // Size of the instructions that are always present in the thunk, 8 byte-aligned.
+    const I_32 extraInstructionsSize = 80;
+    // Size of thunk including the numInstructions generated to store all arguments in memory, 8-byte aligned.
+    metaData->thunkSize = ((( numInstructions * 6 + extraInstructionsSize) / 8) + 1) * 8;
+    thunkMem = (I_8 *)vmFuncs->allocateUpcallThunkMemory(metaData);
+    if (NULL == thunkMem) {
+        return NULL;
+    }
+    metaData->thunkAddress = (void *)thunkMem;
+
+    // Generate the instruction sequence to copy data from C side to Java side.
+    I_32 gprIndex = 1;
+    I_32 maxGPRIndex = 3;
+    I_32 fprIndex = 0;
+    I_32 maxFPRIndex = 6;
+    I_32 currArgIndex = 0;
+
+    // Now we loop through the arguments and generate the actual instructions to store any register arguments in the caller frame.
+    thunkMem += STMG(thunkMem, R6, R15, R4, 1808);
+    thunkMem += LAY(thunkMem, R4, 0, R4, -256);
```

Comment:
Can we use #define to for such constant offset to caller frame instead of hardcoding.

Reply:
Just to make sure, are you referring to the `1808` and `256` constants?

Reply:
Yes, Such stack offset should be more readable.

Reply:
I added new const fields to explain the origination of these two constants.

`256` is the stack frame size and we already have `thunkFrameSize` constant that we can use here.

`1808 = 2048(defaultStackPointerBias) - 256(thunkFrameSize) + 96 (registerSaveAreaSize) - 8 * 10(numRegistersToSave)`

I think this works better as it explains the origination of `1808` especially. It also provides more flexibility to change things in the future. Let me know if you disagree and I'll switch to using `defines` instead. 

The change is in [968f54e](https://github.com/eclipse-openj9/openj9/pull/19701/commits/968f54e09c9943b50bacbc4329189a3ef85ec21f).